Add telemetry to System.Net.NameResolution

MihaZupan opened the pull request:
Added 2 counters:
- `dns-lookups-requested` - total number of lookups since process started*
- `dns-lookups-duration` - EventCounter for min/mean/max duration of requests in the last event counter interval

Added 3 EventSource events:
- `ResolutionStart(hostNameOrAddress)`
- `ResolutionStop()`
- `ResolutionFailed()`

*since telemetry was enabled. This can be changed to _since process started_ by moving the `Interlocked.Increment` calls outside the `IsEnabled()` check. That would avoid the race condition mentioned [for Net.Http](https://github.com/dotnet/runtime/pull/37619#discussion_r443698540) at the cost of an extra Interlocked operation even when telemetry is disabled.

I've manually verified that the counters and events are emitted.

Contributes to #37428

MihaZupan committed:
Add telemetry to System.Net.NameResolution

MihaZupan committed:
Rename hostName to hostNameOrAddress

MihaZupan committed:
Use constants for event IDs

MihaZupan commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+
+
+        [NonEvent]
+        private static Span<char> FormatIPAddressNullTerminated(IPAddress address, Span<char> destination)
```
Do static methods have to be marked with `NonEvent`?

brianrob replied:
You should mark any methods that are not events as `[NonEvent]`, but no, I don't believe that this is strictly required for static methods.

MihaZupan commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
                 ValidateHostName(hostName);
-                return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+
+                if (NameResolutionTelemetry.IsEnabled)
+                {
+                    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
+
+                    Task coreTask;
+                    try
+                    {
+                        coreTask = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                    }
+                    catch when (LogFailure(hostName, stopwatch))
+                    {
+                        Debug.Fail("LogFailure should return false");
+                        throw;
+                    }
+
+                    coreTask.ContinueWith(
+                        (task, state) =>
+                        {
+                            var tuple = (Tuple<string, ValueStopwatch>)state!;
+                            NameResolutionTelemetry.Log.AfterResolution(
+                                hostNameOrAddress: tuple.Item1,
+                                stopwatch: tuple.Item2,
+                                successful: task.IsCompletedSuccessfully);
+                        },
+                        state: Tuple.Create(hostName, stopwatch),
+                        cancellationToken: default,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
+
+                    // coreTask is not actually a base Task, but Task<IPHostEntry> / Task<IPAddress[]>
+                    // We have to return it and not the continuation
+                    return coreTask;
+                }
+                else
+                {
+                    return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                }
```
Alternatively, the PAL itself [could be instrumented](https://gist.github.com/MihaZupan/882eac6ec582ef07dc74fba014011102/revisions) at the cost of increasing the state object by the size of ValueStopwatch (8 bytes)

MihaZupan commented:
/azp run runtime-libraries outerloop

davidsh commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
+        public ValueStopwatch ResolutionStart(string hostNameOrAddress)
```
You should only add an event with a 'Start' suffix if you pair it with an event with a 'Stop' suffix. ActivityId threading will be messed up otherwise. If you don't need ActivityID threading at this level, you should use a different name for the event.

MihaZupan replied:
I renamed the event from `ResolutionStart` to `ResolutionStarting`.

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
+                    // coreTask is not actually a base Task, but Task<IPHostEntry> / Task<IPAddress[]>
+                    // We have to return it and not the continuation
+                    return coreTask;
+                }
```
I'd rather this be a separate async method, e.g.
```C#
return NameResolutionTelemetry.IsEnabled ?
    WithLogging(hostName, justReturnParsedIp, throwOnIPAny, justAddresses) :
    NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);

static async Task WithLogging(string hostName, bool ustAddresses)
{
    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
    bool successful = true;
    try
    {
        await NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
    }
    catch
    {
        LogFailure(hostName, stopwatch);
        successful = false;
    }

    NameResolutionTelemetry.Log.AfterResolution(hostName, stopwatch, successful);
}
```

stephentoub replied:
Ah, I just saw your comment at the end. Ok.

MihaZupan replied:
But `NameResolutionPal.GetAddrInfoAsync` returns a `Task` object that is actually `Task<IPHostEntry>` / `Task<IPAddress[]>`, so we have to return that task itself

noahfalk replied:
Unresolving this to make sure the comment I added is visible, resolve again any time

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
```
Don't we need to check that this event level is actually enabled, e.g. `if (IsEnabled(EventLevel.Informational))` in order to avoid the WriteEvent overhead when it's not?

And once we do that, don't we need to separate the incrementing of the counters from the actual event tracing logic?

cc: @brianrob

MihaZupan replied:
Changed. They are now in the form of
```c#
if (IsEnabled())
{
    Interlocked.Increment(ref _lookupsRequested);

    if (IsEnabled(EventLevel.Informational, EventKeywords.None))
    {
        ResolutionStarting(hostNameOrAddress);
        return ValueStopwatch.StartNew();
    }
}

return default;
```

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
```
There should be a separate void-returning private method that actually does the WriteEvent and that has the right signature for the data being traced.

MihaZupan replied:
Is that really needed? I took inspiration from [AspNetCore](https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionsEventSource.cs#L57-L68)

stephentoub replied:
This pattern is very non-standard.

brianrob replied:
There are a few things that are odd about this event.  In general, you want your events to be of the form:

```
[Event(id,...)]
public void EventName(args...)
{
    WriteEvent(id, args...);
}
```

If the payload arguments are expensive to compute, you can call `EventSource.IsEnabled` before calling the event method.  If you want to encapsulate this all into one method, you can do so by creating a method and marking it as [NonEvent].

It's especially important that when you do define events that their method signature matches the set of arguments that you pass to `WriteEvent`.  If this invariant isn't maintained then the events won't be consumable by analysis tools.  `EventSource` does work to emit metadata about the events, and that metadata is based on the method signature of each event that has the `[Event]` attribute attached to it.

MihaZupan replied:
I assumed that if the signature of the `[Event]` method is string, passing a `ROS<char>` manually would be transparent to EventSource.
Are you saying this event should have a signature of `ROS<char>`, and then have either a string/span passed to it? Are Spans recognized types by EventSource?

brianrob replied:
As of today, `Span` types are not handled by `EventSource`, so you would need to convert it to a type that is handled by `EventSource`, such as `string`, an array, etc.  I think I understand what you're trying to do now - you should just have a `[NonEvent]` method that does the convert from `Span` to `string`, and then have it call the `[Event]` method passing the string.

@noahfalk, I think adding support for `Span` types to be converted into supported logging types would be a worthy thing to do.

MihaZupan replied:
The idea behind all the Span/string logic was precisely to avoid the Span to string conversion.

Is there an issue behind passing a `ROS<char>` in the same format a string would be passed, if the same event ID is used?

brianrob replied:
I see.  I think this will work.  Though, we really should implement this support in the platform so that you don't have to do this.  I'd recommend putting some comments around this so that it's clear what you're doing when someone comes along and finds this in a year or two.

MihaZupan replied:
Changed. The `[Event]` methods are now void-returning and `[NonEvent]` code calls into them when possible.
The `IPAddress` codepath calls into custom `WriteEvent` overloads with the same ID and data format to avoid an extra string allocation.

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, Span<char> arg1, double arg2)
```
Do we need both the above overload and this one?  This one can be made to take a `ReadOnlySpan<char>` instead of a `Span<char>`, and then delete the other one.

MihaZupan replied:
I kept both to differentiate them based on null-termination handling.

I can reuse the ROS<char> one and debug assert that the ROS<char> is null terminated _after_ its end

stephentoub replied:
Why does null termination matter?  The exact length is passed down to the WriteEventCore.

MihaZupan replied:
> The exact length is passed down to the WriteEventCore

For the Span overload yes, as it's ensured that the last character is null. For the string overload `Length + 1` is passed.

stephentoub replied:
@brianrob, given EventData includes the Size, do you know why strings need to be null terminated?

brianrob replied:
The reason is mostly separation of concerns.  The ETW runtime doesn't know the type of the data or the encoding - it just gets a blob and a size and uses the size to write out the appropriate amount of the blob to the event stream.  At this point, the size is thrown away (it is not persisted into the etl).

On the other side, the consumer knows how to decode the event payload.  NULL-terminated strings don't have a size associated with them in their written format - the consumer will just read until it hits a NULL.  There is another string encoding called counted string where the count is pre-pended to the string, and then you don't need the NULL char.  If `EventSource` wrote counted strings, it would be responsible for prepending the count and then specifying the pointer to the count+string as the value assigned to `DataPointer`.  Then `Size` would be the combination of the two.  But again, `Size` would be thrown away and not persisted by the ETW runtime, but the consumer would know it is a counted string and so it would not depend on a NULL char, but would require a count to know how big the string is.

MihaZupan replied:
In that case, shouldn't there at least be a Debug asset in EventSource to catch null chars in string payloads?

brianrob replied:
Unfortunately we can't really do that because we ship release binaries, so by the time you get into `EventSource` code, the asserts have all been stripped out.  The exception to this is for framework eventsources, but that's a pretty small exception.

This is one of the classic issues that has been discussed time and time again, and so far we've not found a nice and pay-for-play way to address this.  It's considered by design that if you pass a string with embedded nulls, you're going to have incorrect decoding behavior in the consumer.  That's the case for both ETW and EventPipe.  I don't know enough about the serialization mechanism to be able to say for sure with LTTng, and I've not tested it.

MihaZupan replied:
Considering that the overloads have different assumptions about the input (Span includes the null character, string has one implicitly at the end), I believe these should remain as separate overloads.

brianrob commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                WriteEvent(
```
Same thing here.  You should have this call the `[Event]` methods directly.  The other benefit of doing this is that in the future, when someone adds a new field to the event, you won't have a mismatch here.

MihaZupan replied:
Based on the conversation above https://github.com/dotnet/runtime/pull/38409#discussion_r447069505, I am marking this as resolved as we can not call into `[Event]` methods directly without the overhead of an extra string allocation.

brianrob commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,214 @@
+
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, Span<char> arg1)
```
These overloads to `WriteEvent` should be safe as long as they are called from an `[Event]` method and the args passed to them match exactly, the signature of the method that calls them.  If they are called from elsewhere or the signature doesn't match, there is no guarantee that the result will be consumable.

MihaZupan replied:
I added a comment about the Span `WriteEvent` overloads mimicking the behavior of string ones.

davidfowl commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
+                        throw;
+                    }
+
+                    coreTask.ContinueWith(
```
Why not use async await?

MihaZupan replied:
Marking as resolved. See comment below https://github.com/dotnet/runtime/pull/38409#discussion_r447076076.

davidfowl commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
+                                stopwatch: tuple.Item2,
+                                successful: task.IsCompletedSuccessfully);
+                        },
+                        state: Tuple.Create(hostName, stopwatch),
```
This allocates, just use async/await instead.

MihaZupan replied:
See [comment](https://github.com/dotnet/runtime/pull/38409/files#diff-434fc33b2da88cc8b47b46b328d07222R557-R559) - the task from `NameResolutionPal.GetAddrInfoAsync` has to be returned directly.

MihaZupan replied:
Marking this as resolved as I do not believe it is possible to implement this with async await since we need control over the returned Task.

As the logic now also checks that the event will be fired (`stopwatch.IsActive`) before attaching the continuation, an async await approach would also have more overhead in the scenario where only counters are enabled.

MihaZupan committed:
Address PR feedback

MihaZupan committed:
Reduce overhead when only the counter is needed

MihaZupan commented:
Gist of new changes:
- Renamed `ResolutionStart` to `ResolutionStarting`
- Made all `[Event]` methods void-returning
- Added the `IsEnabled(EventLevel.Informational)` condition before writing events
- Only calling `ContinueWith` and allocating the state if events would be fired (`IsEnabled(EventLevel.Informational)`)
- Added a comment about `ResolutionSuccess` and `ResolutionFailure` signatures having to match
- Added a comment about the Span `WriteEvent` overloads mimicking the behavior of string ones

noahfalk commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,231 @@
+                if (IsEnabled(EventLevel.Informational, EventKeywords.None))
+                {
+                    ResolutionStarting(hostNameOrAddress);
+                    return ValueStopwatch.StartNew();
```
I'm skeptical that the stopwatch should be conditional on being at the informational level. The usual pattern elsewhere is that counters are tracked as long as the EventSource is enabled. I'd anticipate that if dotnet-counters tool monitored this code it report the value of duration counter is always zero because the EventSource would not have Informational level enabled.

MihaZupan replied:
Good catch, this is my mistake after adding the `IsEnabled(EventLevel.Informational)`, the stopwatch should be returned as long as EventSource is enabled.

noahfalk reviewed (commented):
Great to see more telemetry being added : ) A few questions/recommendations inline

MihaZupan commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,231 @@
+        // Methods below assume that ResolutionSuccess and ResolutionFailure have the same signature
+
+        [Event(ResolutionStartingEventId, Level = EventLevel.Informational)]
+        private void ResolutionStarting(string hostNameOrAddress) => WriteEvent(ResolutionStartingEventId, hostNameOrAddress);
```
I was told that the overhead of tracking the activity that comes from using Start/Stop is not desired in this case.
If it is not significant, I agree a Start/Stop activity would be better here.

alnikola replied:
I believe we need to have activity here as it's required for timings tracing as it's requested in #827.

MihaZupan replied:
Changed them to Start/Stop, which allows us to further simplify the code as we no longer need the hostname/duration for failed/stop events.

noahfalk replied:
Activity tracking does have some small overhead, but it is optional thing that listeners can turn on if they want it (by enabling the TplEventSource keyword 0x80).

MihaZupan commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,48 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
                 ValidateHostName(hostName);
-                return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+
+                if (NameResolutionTelemetry.IsEnabled)
```
In this case, it's explicitly checked before-hand as the `IsEnabled` branch brings more overhead.

I was basing the pattern around Stopwatches on [this gist](https://gist.github.com/anurse/af1859663ac91c6cf69c820cebe92303#event-patterns), where the `Start` method performs the enabled check and avoids the GetTicks overhead when disabled.
Call sites have to keep the Stopwatch timestamp around until the end of the request, so this approach simplifies the call sites by allowing them to keep the same code path regardless of whether telemetry is enabled, at a very low overhead if it's not.

noahfalk replied:
> In this case, it's explicitly checked before-hand as the IsEnabled branch brings more overhead

For comparison, the GetHostName() function also calls BeforeResolution() and AfterResolution() which is extra overhead (albeit small), but it doesn't have the if(Log.IsEnabled) check. I was trying to determine if there was a reason that the different functions use different patterns.

MihaZupan replied:
The overhead in GetHostName is tiny (we could avoid 1 branch?) so I find a shared code path to be simpler.

The overhead in this Task-based method is more significant, as we have to attach a continuation which incurs an allocation as well.

noahfalk replied:
ok, that seems like a fine way to distinguish : ) 👍

MihaZupan committed:
Merge branch 'master' into telemetry-dns

MihaZupan committed:
Make NameResolution Telemetry an activity

ManickaP commented on src/libraries/System.Net.NameResolution/src/System.Net.NameResolution.csproj:
```diff
@@ -9,13 +9,16 @@
     <Compile Include="$(CommonPath)System\Threading\Tasks\TaskToApm.cs"
              Link="Common\System\Threading\Tasks\TaskToApm.cs" />
+    <Compile Include="$(CommonPath)Extensions\ValueStopwatch\ValueStopwatch.cs"
+         Link="Common\Extensions\ValueStopwatch\ValueStopwatch.cs" />
```
Nit: whitespaces.

MihaZupan replied:
Fixed.

ManickaP commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -18,7 +20,21 @@ public static string GetHostName()
+            {
+                name = NameResolutionPal.GetHostName();
+            }
+            catch when (LogFailure(stopwatch))
```
Nice trick, I like it :smile:

ManickaP commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -466,7 +524,46 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
+                    }
+
+                    // Only pay for the overhead if events will be fired
+                    if (stopwatch.IsActive)
```
I guess we check this just because someone could have toggle off the telemetry in between https://github.com/dotnet/runtime/pull/38409/files#diff-434fc33b2da88cc8b47b46b328d07222R528 and here?

MihaZupan replied:
This was a leftover check from an invalid optimization. It is redundant, removed.

ManickaP reviewed (approved):
LGTM, no expert here though, you probably want other approvals as well.

MihaZupan committed:
Remove redundant check

MihaZupan committed:
Fix indentation

alnikola commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -327,20 +354,57 @@ private static object GetHostEntryOrAddressesCore(IPAddress address, bool justAd
+                throw;
+            }
+
+            NameResolutionTelemetry.Log.AfterResolution(stopwatch, successful: true);
```
Please, check if the logging is enabled here at the call site to allow linker trim the call when appropriate. [Take a look at my PR.](https://github.com/dotnet/runtime/blob/abd9c9d468a32945582f2e61cf93d5d363d4efca/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2326)

alnikola commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -0,0 +1,158 @@
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
```
IsEnabled call needs to take the specific event level.

alnikola reviewed (commented):
Overall looks good, but please fix IsEnabled checks.

MihaZupan committed:
Merge master

MihaZupan committed:
Check IsEnabled before AfterResolution to help linker

MihaZupan commented:
/azp run runtime-libraries outerloop

MihaZupan commented:
@brianrob You requested changes, I believe they were addressed. If so, please approve so that the PR can be merged.

brianrob reviewed (approved):
LGTM, thanks!

MihaZupan committed:
Add Telemetry tests

MihaZupan committed:
Merge master

MihaZupan commented:
/azp run runtime-libraries outerloop

MihaZupan committed:
Throws => ThrowsAny

MihaZupan commented:
/azp run runtime-libraries outerloop

MihaZupan commented:
Failure: #37540
